harness/determined/searcher/_search_runner.py

```diff
@@ -0,0 +1,351 @@
+                            ):
+                                self.state.experiment_completed = True
+
+                            if (
+                                event.experimentInactive.experimentState
+                                == bindings.determinedexperimentv1State.STATE_PAUSED
+                            ):
+                                self._show_experiment_paused_msg()
+                            else:
+                                experiment_is_active = False
+                            break
+
+                        else:
+                            operations = self._get_operations(event)
+
+                        # save state
+                        self.state.last_event_id = event.id
+                        self.save_state(experiment_id, operations)
+
+                    first_event = False
+                    self.post_operations(session, experiment_id, event, operations)
+
+        except KeyboardInterrupt:
+            print("Runner interrupted")
+
+    def post_operations(
+        self,
+        session: client.Session,
+        experiment_id: int,
+        event: bindings.v1SearcherEvent,
+        operations: List[searcher.Operation],
+    ) -> None:
+        body = bindings.v1PostSearcherOperationsRequest(
+            experimentId=self.state.experiment_id,
+            searcherOperations=[op._to_searcher_operation() for op in operations],
+            triggeredByEvent=event,
+        )
+        bindings.post_PostSearcherOperations(
+            session,
+            body=body,
+            experimentId=experiment_id,
+        )
+
+    def get_events(
+        self,
+        session: client.Session,
+        experiment_id: int,
+    ) -> Optional[Sequence[bindings.v1SearcherEvent]]:
+        # API is implemented with long polling.
+        events = bindings.get_GetSearcherEvents(session, experimentId=experiment_id)
+        return events.searcherEvents
+
+    def save_state(self, experiment_id: int, operations: List[searcher.Operation]) -> None:
+        pass
+
+    def _show_experiment_paused_msg(self) -> None:
+        pass
+
+    @staticmethod
+    def _searcher_event_as_dict(event: bindings.v1SearcherEvent) -> dict:
+        d = {}
+        if event.trialExitedEarly:
+            d["trialExitedEarly"] = event.trialExitedEarly.to_json()
+        if event.validationCompleted:
+            d["validationCompleted"] = event.validationCompleted.to_json()
+        if event.trialProgress:
+            d["trialProgress"] = event.trialProgress.to_json()
+        if event.trialClosed:
+            d["trialClosed"] = event.trialClosed.to_json()
+        if event.trialCreated:
+            d["trialCreated"] = event.trialCreated.to_json()
+        if event.initialOperations:
+            d["initialOperations"] = event.initialOperations.to_json()
+        if event.experimentInactive:
+            d["experimentInactive"] = event.experimentInactive.to_json()
+        d["id"] = event.id
+        return d
```

Comment:
is this function materially different than:

```python
return {k: v for k, v in event.to_json() if v is not None}
```

It might be nice to lean on the generated bindings here for their type safety.